packages/langium/src/grammar/typification/type-validator.ts

```diff
@@ -0,0 +1,142 @@
+}
+
+function collectValidationResources(grammar: Grammar): ValidationResources {
+    const astResources = collectAllAstResources([grammar]);
+    const inferred = collectInferredTypes(Array.from(astResources.parserRules), Array.from(astResources.datatypeRules));
+    const declared = collectDeclaredTypes(Array.from(astResources.interfaces), Array.from(astResources.types), inferred);
+
+    const nameToDeclaredType = mergeTypesAndInterfaces(declared)
+        .reduce((acc, type) => acc.set(type.name, type), new Map<string, Type>());
+
+    const typeNameToRule = grammar.rules
+        .reduce((acc, rule) => acc.add(getRuleType(rule), rule),
+            new MultiMap<string, AbstractRule>()
+        );
+    const nameToInferredType = mergeTypesAndInterfaces(inferred)
+        .reduce((acc, type) => acc.set(type.name, { type, rules: typeNameToRule.get(type.name) }),
+            new Map<string, TypeToRules>()
+        );
+
+    return { nameToDeclaredType, nameToInferredType };
+}
+
+function mergeTypesAndInterfaces(astTypes: AstTypes): Type[] {
+    return (astTypes.interfaces as Type[]).concat(astTypes.types);
+}
+
+function checkStringElementConsistency(inferredType: string[], declaredType: string[], errorElementName: string): string[] {
+    const extra = inferredType.filter(e => !declaredType.includes(e));
+    const lack = declaredType.filter(e => !inferredType.includes(e));
+    return getExtraAndLackError(extra, lack, errorElementName);
+}
+
+function checkAlternativesConsistency(inferredType: FieldType[], declaredType: FieldType[], errorElementName = 'type alternative'): string[] {
+    return checkStringElementConsistency(inferredType.map(typeFieldToString), declaredType.map(typeFieldToString), errorElementName);
+}
+
+type ComparingFields = {
+    inferredField: Field;
+    declaredField: Field;
+}
+
+function checkFieldConsistency(inferredType: Field[], declaredType: Field[]): string[] {
+    const inconsistencyReasons: string[] = [];
+    const commonByName: ComparingFields[] = [];
+    const extra: string[] = [];
+    for (const inferredField of inferredType) {
+        const declaredField = declaredType.find(e => e.name === inferredField.name);
+        if (declaredField) {
+            commonByName.push({ inferredField, declaredField });
+        } else {
+            extra.push(inferredField.name);
+        }
+    }
+
+    const lack: string[] = declaredType
+        .filter(declaredField => !inferredType.some(inferredField => inferredField.name === declaredField.name))
+        .map(e => e.name);
+    inconsistencyReasons.push(...getExtraAndLackError(extra, lack, 'field'));
+
+    for (const comparingFields of commonByName) {
+        const inferred = comparingFields.inferredField;
+        const declared = comparingFields.declaredField;
+        inconsistencyReasons.push(...checkAlternativesConsistency(inferred.typeAlternatives, declared.typeAlternatives, `type alternative of '${inferred.name}' field`));
```

Comment:
I see two errors when a wrong type is assigned (e.g. declared number, but assigned string):
> There is no 'string' type alternative of 'name' field in declared type.
> Lack of 'number' type alternative of 'name' field in inferred type.

I propose to show a single message like this instead:
> The assigned type 'string' is not compatible with the declared property 'name' of type 'number'.

In general, please use the term "property" (JavaScript) instead of "field" (Java).